### PR TITLE
Adopt V2 throughout the solidity/typescript repo

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR transitions the non-agent code to V2. It removes vestigial contracts, removes the "V2" suffix where present, and updates the SDK and infra code to work with V2.

### CODE REVIEW

File: solidity/contracts/InterchainGasPaymaster.sol [hunk 1]
```diff
@@ -39,20 +38,19 @@ contract InterchainGasPaymaster is IInterchainGasPaymaster, OwnableUpgradeable {
      * @notice Deposits msg.value as a payment for the relaying of a message
      * to its destination chain.
      * @param _outbox The address of the Outbox contract.
-     * @param _leafIndex The index of the message in the Outbox merkle tree.
      * @param _destinationDomain The domain of the message's destination chain.
      */
     function payGasFor(
         address _outbox,
-        uint256 _leafIndex,
+        bytes32 _messageId,
```
Comment (asaj) [inline solidity/contracts/InterchainGasPaymaster.sol:45]: Should we remove `_outbox` from the signature here?

File: typescript/utils/src/utils.ts [hunk 2]
```diff
@@ -74,95 +101,29 @@ export const formatMultisigModuleMetadata = (
...
  *
  * @param message
  * @returns
  */
 export function parseMessage(message: string): ParsedMessage {
   const buf = Buffer.from(utils.arrayify(message));
-  const origin = buf.readUInt32BE(0);
-  const sender = utils.hexlify(buf.slice(4, 36));
-  const destination = buf.readUInt32BE(36);
+  const version = buf.readUInt32BE(0);
```
Comment (asaj) [inline typescript/utils/src/utils.ts:112]: This is now a uint8

File: typescript/utils/src/utils.ts [hunk 3]
```diff
@@ -74,95 +101,29 @@ export const formatMultisigModuleMetadata = (
...
+  const nonce = BigNumber.from(utils.hexlify(buf.slice(4, 36)));
+  const origin = buf.readUInt32BE(36);
+  const sender = utils.hexlify(buf.slice(40, 72));
+  const destination = buf.readUInt32BE(72);
```
Comment (asaj) [inline typescript/utils/src/utils.ts:116]: Recipient indices are wrong

**Comment by asaj** [comment]
e2e tests are expected to fail

File: typescript/infra/config/environments/mainnet/helloworld/abacus/addresses.json [hunk 4]
```diff
@@ -1,26 +0,0 @@
-{
```
Comment (tkporter) [inline typescript/infra/config/environments/mainnet/helloworld/abacus/addresses.json:1]: why delete all these types of files? bc we'll need to redeploy them for v2?

File: typescript/infra/scripts/debug-message.ts [hunk 5]
```diff
@@ -66,7 +66,8 @@ async function checkMessage(
   multiProvider: MultiProvider<any>,
   message: DispatchedMessage,
 ) {
-  console.log(`Leaf index: ${message.leafIndex.toString()}`);
+  const messageId = utils.messageId(message.message);
+  console.log(`Id: ${messageId}`);
```
Comment (tkporter) [inline typescript/infra/scripts/debug-message.ts:70]: I think still logging the nonce (i.e. leaf index) could be useful

File: typescript/infra/src/core/deploy.ts [hunk 6]
```diff
@@ -47,38 +54,10 @@ export class HyperlaneCoreInfraDeployer<
...
-              validatorManager: inboxContracts?.inboxValidatorManager.address,
-            },
-          } as const;
-
-          rustConfig.inboxes[remote] = inbox;
-        });
-      writeJSON(directory, `${chain}_config.json`, rustConfig);
+      rustConfig.chains[chain] = chainConfig;
     });
+    writeJSON(directory, 'rust_config.json', rustConfig);
```
Comment (tkporter) [inline typescript/infra/src/core/deploy.ts:61]: I take it the accompanying v2 agent PR requires a config like this? (I haven't taken a look yet)

File: typescript/sdk/src/core/testHyperlaneDeploy.hardhat-test.ts [hunk 7]
```diff
@@ -21,8 +21,8 @@ const message = '0xdeadbeef';
 
 describe('TestCoreDeployer', async () => {
   let testCoreApp: TestCoreApp,
-    localOutbox: TestOutbox,
-    remoteOutbox: TestOutbox,
+    localOutbox: TestMailbox,
```
Comment (tkporter) [inline typescript/sdk/src/core/testHyperlaneDeploy.hardhat-test.ts:24]: wondering if this should be `localMailbox` and similar for the next line

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 8]
```diff
@@ -46,208 +34,41 @@ export class HyperlaneCoreChecker<
...
         contracts.upgradeBeaconController,
-        contracts.outbox.contract,
-        contracts.outboxValidatorManager,
-        ...Object.values(contracts.inboxes)
-          .map((inbox: any) => [
-            inbox.inbox.contract,
-            inbox.inboxValidatorManager,
-          ])
-          .flat(),
+        // contracts.mailbox.contract,
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:38]: uncomment?

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts [hunk 9]
```diff
@@ -34,10 +24,8 @@ export class HyperlaneCoreChecker<
 
     await this.checkDomainOwnership(chain);
     await this.checkProxiedContracts(chain);
-    await this.checkOutbox(chain);
-    await this.checkInboxes(chain);
-    await this.checkConnectionManager(chain);
-    await this.checkValidatorManagers(chain);
+    await this.checkMailbox(chain);
+    // await this.checkDefaultIsm(chain);
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:28]: does this exist yet? should `checkValidatorManager` be changed into this?

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 10]
```diff
@@ -42,77 +38,62 @@ export class HyperlaneCoreDeployer<
...
-    partialDeployment?: Partial<CoreContractsMap<Chain>>,
-  ): Promise<CoreContractsMap<Chain>> {
-    return super.deploy(partialDeployment) as Promise<CoreContractsMap<Chain>>;
-  }
-
-  async deployOutbox<LocalChain extends Chain>(
+  async deployMailbox<LocalChain extends Chain>(
     chain: LocalChain,
-    config: ValidatorManagerConfig,
+    moduleAddress: types.Address,
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:43]: maybe `defaultIsm`?

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 11]
```diff
@@ -42,77 +38,62 @@ export class HyperlaneCoreDeployer<
...
+  ): Promise<MultisigIsm> {
+    const module = await this.deployContract(chain, 'multisigIsm', []);
+    const configChains = Object.keys(this.configMap) as Chain[];
+    const remotes = this.multiProvider
+      .intersect(configChains, false)
+      .multiProvider.remoteChains(chain);
+    await super.runIfOwner(chain, module, async () => {
+      // TODO: Remove extraneous validators
+      for (const remote of remotes) {
+        const moduleConfig = this.configMap[remote].multisigIsm;
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:69]: also wondering how you'd feel about `s/module/ism/g` throughout here

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 12]
```diff
@@ -137,72 +118,19 @@ export class HyperlaneCoreDeployer<
...
-    );
+    const multisigIsm = await this.deployMultisigIsm(chain);
 
-    const outbox = await this.deployOutbox(
+    const mailbox = await this.deployMailbox(
       chain,
-      config.validatorManager,
+      multisigIsm.address,
       upgradeBeaconController.address,
     );
```
Comment (tkporter) [inline typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:127]: look at all that deleted code 🤤

**Review by tkporter — APPROVED** [review]
some small things but generally lgtm! I'll also disclaim I've never been super close to the SDK so it's possible I've missed something there, hoping @yorhodes or @nambrot can give a more confident review there

File: solidity/test/mailboxv2.test.ts [hunk 13]
```diff
@@ -12,8 +12,8 @@ import {
   BadRecipient6__factory,
   TestIsm,
   TestIsm__factory,
-  TestMailboxV2,
-  TestMailboxV2__factory,
+  TestMailbox,
+  TestMailbox__factory,
   TestRecipient__factory,
 } from '../types';
```
[the remark this answers is not in the document]
Comment (tkporter) [reply solidity/test/mailboxv2.test.ts:19]: I take this back, now seeing https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1198

File: solidity/contracts/InterchainGasPaymaster.sol [hunk 14]
```diff
@@ -16,11 +16,10 @@ contract InterchainGasPaymaster is IInterchainGasPaymaster, OwnableUpgradeable {
 
     /**
      * @notice Emitted when a payment is made for a message's gas costs.
-     * @param outbox The address of the Outbox contract.
-     * @param leafIndex The index of the message in the Outbox merkle tree.
+     * @param messageId The ID of the message to pay for.
      * @param amount The amount of native tokens paid.
      */
-    event GasPayment(address indexed outbox, uint256 leafIndex, uint256 amount);
+    event GasPayment(bytes32 messageId, uint256 amount);
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/contracts/InterchainGasPaymaster.sol:22]: Yes definitely, good catch

File: solidity/contracts/test/TestMailboxV2.sol [hunk 15]
```diff
@@ -22,4 +26,17 @@ contract TestMailboxV2 is MailboxV2 {
         }
         return _proof;
     }
+
+    function testHandle(
+        uint32 origin,
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/contracts/test/TestMailboxV2.sol:31]: 100%

File: solidity/test/mailboxv2.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/test/mailboxv2.test.ts:19]: Yeah it helped keep the diff cleanest to do things this way

File: typescript/infra/config/environments/mainnet/helloworld/abacus/addresses.json (hunk 4 above)
Comment (asaj) [reply typescript/infra/config/environments/mainnet/helloworld/abacus/addresses.json:1]: Yes exactly

File: typescript/infra/src/core/deploy.ts (hunk 6 above)
Comment (asaj) [reply typescript/infra/src/core/deploy.ts:61]: It does indeed

File: typescript/infra/scripts/debug-message.ts (hunk 5 above)
Comment (asaj) [reply typescript/infra/scripts/debug-message.ts:70]: We do this when we log `message.parsed`

File: typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts (hunk 9 above)
Comment (asaj) [reply typescript/sdk/src/deploy/core/HyperlaneCoreChecker.ts:28]: Yes it should have been, good catch

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts (hunk 11 above)
Comment (asaj) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:69]: Good catch, missed this.

**Pull request merged**